5.x: add phpdoc generics

LordSimal opened the pull request:
I am happy to get some help here 😄 Still have much to learn about generics with PHPDoc ^^

LordSimal commented on src/Database/Connection.php:
```diff
@@ -125,23 +125,25 @@ public function __construct(array $config)
         if (!empty($config['log'])) {
             $this->enableQueryLogging((bool)$config['log']);
         }
+         */
```
This is broken because query logging is not present inside the connection anymore

markstory committed:
Fix more phpstan errors.

markstory commented:
I fixed a bunch of the type errors. I got stuck on resolving the generic parameter types in methods provided by the `EventDispatcherTrait`. The interface types are resolvable with `@implements` but I can't find the equivalent annotation for trait methods.

ADmad commented:
The introduction of generics template for `EventInterface` and `ResultSetInterface` has caused a cascading effect where phpstan expects the specific type to be declared throughout the usage chain. Dealing with all that in a single go will be very time consuming.

I have opened an alternate PR #16892 which disables phpstan generic type check for now and we can deal with them in smaller chunks over time.

LordSimal commented:
Skipping the generic type check seems to be the better approach right now 😅

markstory commented:
> The introduction of generics template for EventInterface and ResultSetInterface has caused a cascading effect where phpstan expects the specific type to be declared throughout the usage chain. Dealing with all that in a single go will be very time consuming.

Yep. I got most of the generics resolved but couldn't get the traits to behave :cry:

LordSimal committed:
rebase

LordSimal committed:
fix duplicate InvalidArgumentException

LordSimal commented:
Just rebased this branch and fixed the duplicate InvalidArgumentException in the ExceptionTrap.
The main remaining part here is definitely the EventDispatcherTrait already mentioned above.
Thanks for the help @markstory 🙇🏻

ADmad commented on src/Console/CommandRunner.php:
```diff
@@ -33,6 +33,9 @@
  * Run CLI commands for the provided application.
+ *
+ * @implements \Cake\Event\EventDispatcherInterface<\Cake\Core\ConsoleApplicationInterface>
+ * @template-uses \Cake\Event\EventDispatcherInterface<\Cake\Core\ConsoleApplicationInterface>
```
What does `@template-uses` do? In psalm's doc I only see `@template-use` which is mean to be used with `use` statement inside classes for traits.

markstory replied:
I don't think it actually does anything. I'll remove it.

ADmad commented on src/Controller/Component/CheckHttpCacheComponent.php:
```diff
@@ -36,7 +36,8 @@ class CheckHttpCacheComponent extends Component
      *
-     * @param \Cake\Event\EventInterface $event The Controller.beforeRender event.
+     * @template TSubject of object
+     * @param \Cake\Event\EventInterface<TSubject> $event The Controller.beforeRender event.
```
```suggestion
     * @param \Cake\Event\EventInterface<\Cake\Controller\Controller> $event The Controller.beforeRender event.
```

We know the type here, in the controller layer the subject for all the events is a `Controller` instance. Similarly in tables and behaviors it would be `Cake\ORM\Table`.

LordSimal committed:
specify controller and table generics in components and behaviors

markstory committed:
Fix more annotation types

dereuromark commented:
If this gets merged and solves the phpstan issue, could we also backport this to 4.5?
Then I could remove all those additional config lines I had to add to make phpstan green again :)

LordSimal committed:
add use phpdoc to EventDispatcherTrait usages

LordSimal committed:
fix cs

LordSimal commented:
I found out via https://github.com/vimeo/psalm/issues/1072 that traits need a `@use` phpdoc.
This now resolves at least the EventDispatcherTrait usages as far as I can tell.

LordSimal committed:
add more generics

LordSimal committed:
adjust phpdoc

LordSimal committed:
regnerate stan baseline

LordSimal commented:
This is as far as I can take it. I regenerated the baselines for phpstan and psalm so CI is happy.
